Makefile

```diff
@@ -0,0 +1,88 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+# This Makefile serves to generate language-specific messages from .proto file.
+
+# we want bash as shell
+SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
+         else if [ -x BASH_PATH="$(command -v bash)" ]; then echo $$BASH_PATH; \
+         else echo sh; fi; fi)
+
+# Set O variable if not already done on the command line;
+# or avoid confusing packages that can use the O=<dir> syntax for out-of-tree
+# build by preventing it from being forwarded to sub-make calls.
+ifneq ("$(origin O)", "command line")
+O := $(CURDIR)/output
+endif
+
+# Remove the trailing '/.' from $(O) as it can be added by the makefile wrapper
+# installed in the $(O) directory.
+# Also remove the trailing '/' the user can set when on the command line.
+override O := $(patsubst %/,%,$(patsubst %.,%,$(O)))
+# Make sure $(O) actually exists before calling realpath on it; this is to
+# avoid empty CANONICAL_O in case on non-existing entry.
+CANONICAL_O := $(shell mkdir -p $(O) >/dev/null 2>&1)$(realpath $(O))
+
+CANONICAL_CURDIR = $(realpath $(CURDIR))
+
+PROTO_DIR = $(CANONICAL_CURDIR)/proto
+RUST_LANG_DIR = $(CANONICAL_CURDIR)/rust
+
+BASE_DIR := $(CANONICAL_O)
+$(if $(BASE_DIR),, $(error output directory "$(O)" does not exist))
+
+# download-location
+DL_DIR := $(shell mkdir -p $(BASE_DIR)/dl >/dev/null 2>&1)$(BASE_DIR)/dl
+
+BUILD_DIR := $(BASE_DIR)/build
+RUST_BUILD_DIR := $(BUILD_DIR)/rust
+
+FILES=$(wildcard proto/edgehog/device/forwarder/*.proto)
+
+PROTOC_CHECK_SCRIPT=$(CANONICAL_CURDIR)/scripts/protoc_check.sh
+
+RUST_LANG=$(RUST_BUILD_DIR)/edgehog-device-forwarder-proto
+RUST_CODEGEN_SCRIPT=$(CANONICAL_CURDIR)/scripts/rust_codegen.sh
+
+# This is our default rule, so must come first
+.PHONY: all
+all: $(RUST_LANG)
+$(RUST_LANG): $(FILES) $(RUST_CODEGEN_SCRIPT)
+		mkdir -p $(RUST_BUILD_DIR)
+		$(RUST_CODEGEN_SCRIPT) codegen $(PROTO_DIR) $(RUST_LANG_DIR) $(RUST_BUILD_DIR)
+
+.PHONY: protoc-check
+protoc-check: $(PROTOC_CHECK_SCRIPT)
+		$(PROTOC_CHECK_SCRIPT)
+
+.PHONY: rust
+rust: protoc-check $(RUST_LANG)
```

Comment:
Is there a reason to use a different version?

The check is necessary to ensure that the established version of Protoc is used while committing, it doesn't make sense to permit the use of a different version here.

Warnings wouldn't help, since they can be ignored.

Reply:
there are 4 possible outcomes when compiling with a different protoc version
1. it compiles the code as expected
2. it crashes, in which case the `make` command fails
3. it compiles successfully but the code it generates is not the same as with `$PROTOC_VERSION`

for 2., the warning will point to the possible cause of the failure
for 3., the CI check will fail anyway

I don't think it's necessary to force the user to use the version we want, but we should inform them of the _officially supported_ version

Reply:
This script is _for us and nobody will ever run it_. It's more beneficial to make the build brake on an incorrect version than warn about it.

Even if somebody runs this they will be aware of the wrong version and download the appropriate version or comment out the check.

> Is there a reason to use a different version?

Also, to drive the point home, the build should break if the proto version is wrong. Since there is no reason that _we_ would want to commit files generated with a different version of `protoc`.

Maybe a compromise would be to have an environment variable to pass this check, but I believe there should be an actual reason to use a different version of protoc.

Reply:
This already happened when we updated from proto 24.3 → 24.4, having the protoc check prevents the user from uploading the wrong code and having to wait for CI to fail with just the diffs being incompitible.